reader.go

```diff
@@ -50,11 +50,23 @@ func NewReader(r io.Reader) *Reader {
 }
 
 // NewReaderSize returns a new *Reader that
-// reads from 'r' and has a buffer size 'n'
+// reads from 'r' and has a buffer size 'n'.
 func NewReaderSize(r io.Reader, n int) *Reader {
+	buf := make([]byte, 0, max(n, minReaderSize))
+	return NewReaderBuf(r, buf)
+}
+
+// NewReaderBuf returns a new *Reader that
+// reads from 'r' and uses 'buf' as a buffer.
+// It panics if the length of the buffer is smaller
+// than 'minReaderSize'.
+func NewReaderBuf(r io.Reader, buf []byte) *Reader {
+	if len(buf) < minReaderSize {
```

Comment:
You'd want `cap(buf)` in this check, not `len(buf)`. But I think it `minReaderSize` would also need to be exported in order for it to be reasonable to ask users to obey this API contract.

Conversely, it might be simpler to just allocate a new buffer in the unusual circumstance that the buffer is too small.

Reply:
That's a good point. I changed:
* Checking `cap` instead of `len`
* Removed panic and allocated minimal buffer instead (also updated the function comment)
* "emptying" the buffer so it will not appear as it has something already buffered
* Added small tests to check that everything works